src/python/pants/util/frozendict.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Any, Dict, Iterable, Iterator, Mapping, Optional, Tuple, TypeVar, Union
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+class FrozenDict(Mapping[K, V]):
+    """A wrapper around a normal `dict` that removes all methods to mutate the instance and that
+    implements __hash__.
+
+    This should be used instead of normal dicts when working with the engine because normal dicts
+    are not safe to use.
+    """
+
+    def __init__(self, item: Optional[Union[Mapping[K, V], Iterable[Tuple[K, V]]]] = None) -> None:
+        """Creates a `FrozenDict` from a mapping object or a sequence of tuples representing
+        entries.
+
+        These values must be immutable and hashable, which we proactively validate.
```

Comment:
Well, you can perfectly well have a mutable object that's hashable - the mutable part just shouldn't (but could be :/) part of the hash. It might be best to just drop immutable here - you're not guaranteeing that.